Speed up dumping and fix some memory issues

patrickbr opened the pull request:
This PR

* uses an updated libutil, in which WKTs are generated by using the `fmt` library for (much) faster `double` to `string` conversion
* removes unnecessary string copying by introducing`writeLiteralTripleUnsafe()`. This in particular avoids copying of all WKT strings just to wrap them in quotes and add "^^wktLiteral".
* introduces a method `IREREFUnsafe` in `Writer`, which is faster than `IREREF`, but is only to be called with safe strings.
* removes all usages of `std::ostringstream` at critical places
* does *not* store ways which are only members of area relations in `RelationHandler`. These are only needed to build their geometries, and we do not build relation geometries for areas (these are completely handled by `libosmium`)
* moves the scopes around a bit so that everything `libosmium`-related should be destroyed by the time the spatial relations are build.

On my machine, this speeds the triple writing up by a factor of 4 on the Malta dataset.

patrickbr commented:
@lehmann-4178656ch, I fixed the tests, everything should run through now. I also experimented with dropping the QLEVER output format. However, with the TTL format, `riot` complains that there are now spaced in the IRIs:

    14:18:58 WARN  riot            :: [line: 534, col: 40] Bad IRI: <https://en.wikipedia.org/wiki/Victoria, Gozo> Spaces are not legal in URIs/IRIs.

The corresponding line in the `.ttl` looks as follows:

    osmnode:129953609 osm2rdfkey:wikipedia <https://en.wikipedia.org/wiki/Victoria,\u0020Gozo> .

I am not sure what this means. Is the `ttl` format implemented incorrectly, or is `riot`'s validation broken? I reverted the changes until we have clarified this. `QLEVER` is now again the default output format again.

lehmann-4178656ch commented:
[Section 6.4](https://www.w3.org/TR/turtle/#sec-escapes) explicitly states that for IRIs numeric escapes (`UCHAR`) shall be used.

> There are three forms of escapes used in turtle documents:
> * numeric escape sequences represent Unicode code points:
>  [...]
> * string escape sequences represent the characters traditionally escaped in string literals:
>  [...]
> * reserved character escape sequences consist of a `'\'` followed by one of `~.-!$&'()*+,;=/?#@%_` and represent the character to the right of the `'\'`.
>  [...]

These do not allow `%-encoding` but... there is a note:

> Note
>
> %-encoded sequences are in the [character range for IRIs](https://www.w3.org/TR/turtle/#grammar-production-IRIREF) and are [explicitly allowed](https://www.w3.org/TR/turtle/#grammar-production-PERCENT) in local names. These appear as a '%' followed by two hex characters and represent that same sequence of three characters. These sequences are not decoded during processing. A term written as `<http://a.example/%66oo-bar>` in Turtle designates the IRI `http://a.example/%66oo-bar` and not IRI `http://a.example/foo-bar`. A term written as `ex:%66oo-bar` with a prefix `@prefix ex: <http://a.example/>` also designates the IRI `http://a.example/%66oo-bar`.

If I read this correctly, %-encoding is allowed but has a different interpretation:

IRIs using `'\u' hex hex hex hex` or `'\U' hex hex hex hex hex hex hex hex` are resolved as show by the error message of riot:
`<https://en.wikipedia.org/wiki/Victoria,\u0020Gozo>` -> `<https://en.wikipedia.org/wiki/Victoria, Gozo>`

Percent encoding would not get resolved:
`<https://en.wikipedia.org/wiki/Victoria,%20Gozo>` -> `<https://en.wikipedia.org/wiki/Victoria,%20Gozo>`

Looking at apache/jena#2626 the possibility exists that we produce well formed RDF/XML as "[...]  or Jena would have produced a parser error"

The relevant code should be https://github.com/apache/jena/blob/main/jena-arq/src/main/java/org/apache/jena/riot/tokens/TokenizerText.java#L528-L579 and [`SPC`](https://github.com/apache/jena/blob/a2692a8cceaef678b73819a0d74dfc215a27b713/jena-base/src/main/java/org/apache/jena/atlas/lib/Chars.java#L192) which uses https://github.com/apache/jena/blob/main/jena-base/src/main/java/org/apache/jena/atlas/io/PeekReader.java . I've not found the place yet where the conversion of `UCHAR` happens as the comment above `readIRI()` explicitly references the same grammar rule we use.

patrickbr commented:
I added `gzip` as an additional compression method, with a hardcoded compression level of 1 for maximum speed. Should result in compression factors around 12 (bz2 has a compression factor around 17). I removed the option `--output-no-compress` and added the option `--output-compression` with valid values `none`, `bz2` (default), and `gz`.

lehmann-4178656ch commented on CMakeLists.txt:
```diff
@@ -78,19 +78,19 @@ set(INSTALL_GMOCK OFF)
 set(BENCHMARK_ENABLE_INSTALL OFF)
 
-find_package(Git QUIET)
+# find_package(Git QUIET)
```
Why keep this as a comment if not needed?

patrickbr replied:
This was actually an oversight, I commented this out for debugging purposes.

lehmann-4178656ch commented on src/osm/CountHandler.cpp:
```diff
@@ -27,23 +27,25 @@ void osm2rdf::osm::CountHandler::prepare_for_lookup() {
 void osm2rdf::osm::CountHandler::node(const osmium::Node& node){
+  if (node.positive_id() < _minId) _minId = node.positive_id();
+  if (node.positive_id() > _maxId) _maxId = node.positive_id();
   if (_firstPassDone || node.tags().empty())  {
```
How many nodes would removing this check add?

patrickbr commented on src/osm/OsmiumHandler.cpp:
```diff
@@ -198,29 +186,31 @@ void osm2rdf::osm::OsmiumHandler<W>::area(const osmium::Area& area) {
 template <typename W>
 void osm2rdf::osm::OsmiumHandler<W>::node(const osmium::Node& node) {
   _nodesSeen++;
+  if (node.tags().empty()) {
```
I think @hannahbast already has a local branch with an option to add all nodes (which disables exactly this check). We should investigate this in a new PR based on this.

patrickbr commented on tests/osm/TagList.cpp:
```diff
@@ -33,15 +33,15 @@ TEST(OSM_TagList, convertTagList) {
   osmium::builder::add_node(
       osmiumBuffer, osmium::builder::attr::_id(42),
       osmium::builder::attr::_location(osmium::Location(7.51, 48.0)),
-      osmium::builder::attr::_tag("city", "Freiburg"),
```
Generally no. OSM allows duplicate tags. The new code not only has a faster translation from osmium-space into our space, but will also keep these duplicate attributes (which might be useful for debugging OSM).
